forecasting/src/autogluon/forecasting/utils/dataset_utils.py

```diff
@@ -0,0 +1,227 @@
+    freq = pd.infer_freq(time_list)
+    if freq is None:
+        raise ValueError("Freq cannot be inferred. Check your dataset.")
+
+    def reshape_dataframe(df):
+        """
+        for each time occurs in the dataset, we select the target value corresponding to
+        each time series, and use dataframe.pivot() to convert it to one column, where the column name is the
+        time, each row is the corresponding target value for each time series.
+        """
+        df = df.sort_values(by=index_column)
+        data_dic = {index_column: sorted(list(set(df[index_column])))}
+
+        for time in time_list:
+            tmp = df[df[time_column] == time][[index_column, time_column, target_column]]
+            tmp = tmp.pivot(index=index_column, columns=time_column, values=target_column)
+            tmp_values = tmp[time].values
+            data_dic[time] = tmp_values
+        return pd.DataFrame(data_dic)
+
+    X = reshape_dataframe(X)
+    return X
+
+
+def train_test_split_dataframe(data, prediction_length):
+    test_ds = data.copy()
+    train_ds = data.iloc[:, :-prediction_length]
+    return train_ds, test_ds
+
+
+def train_test_split_gluonts(data, prediction_length, freq=None):
+    train_data_lst = []
+    test_data_lst = []
+
+    if freq is None:
+        if isinstance(data, ListDataset):
+            freq = data.list_data[0]["start"].freq
+        else:
+            raise ValueError("Cannot infer freq if the data is not ListDataset.")
+
+    for entry in data:
+        test_data_lst.append(entry.copy())
+        tmp = {}
+        for k, v in entry.items():
+            if "dynamic" in k.lower():
+                tmp[k] = v[..., : -prediction_length]
+            elif "target" in k.lower():
+                tmp[k] = v[..., : -prediction_length]
+            else:
+                tmp[k] = deepcopy(v)
+        train_data_lst.append(tmp)
+    train_ds = ListDataset(train_data_lst, freq=freq)
+    test_ds = ListDataset(test_data_lst, freq=freq)
+    return train_ds, test_ds
+
+
+def time_series_dataset(data,
+                        index_column=None,
+                        target_column="target",
+                        time_column="date",
+                        chosen_ts=None,
+                        static_features=None):
+    rebuilt_data = rebuild_tabular(data,
+                                   index_column=index_column,
+                                   target_column=target_column,
+                                   time_column=time_column)
+    if index_column is None:
+        index_column = "index_column"
+    if chosen_ts is not None:
+        rebuilt_data = rebuilt_data.loc[rebuilt_data[index_column].isin(chosen_ts)]
+    return TimeSeriesDataset(rebuilt_data, index_column=index_column, static_features=static_features)
+
+
+def extract_static_feature(index_column, features):
+    #TODO: Find a way to do fillna
+    if index_column is None:
+        index_column = "index_column"
```

Comment:
Should we raise error if there are static features but no index column?

Reply:
We should also check that the index column values match  the index values in static_features and raise error if not.